File: akka-samples/scala-samples/src/main/scala/actors/future/ReceiveFuture.scala
package actors.future

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorSystem, Behavior}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.{Failure, Success}

object ReceiveFuture {
  def main(args: Array[String]): Unit = {
    ActorSystem[Int](guardian(), "receive-future")
  }

  def compute(): Future[Int] = {
    Future {
      Thread.sleep(3 * 1000)
      42
    }
  }

  def guardian(): Behavior[Int] =
    Behaviors.setup[Int] { context =>
      println("Starting...")

      context.pipeToSelf(compute()) {
        case Success(value) => value
        case Failure(_) => -1
      }

      Behaviors.receive[Int] { (_, message) =>
        println(message)
        Behaviors.stopped
      }
    }
}

File: akka-samples/java-samples/src/main/java/actors/future/ReceiveFuture.java
package actors.future;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

import java.util.concurrent.*;

public class ReceiveFuture extends AbstractBehavior<Integer> {
    public CompletableFuture<Integer> calculate() {
        return CompletableFuture.supplyAsync(() -> {
            try { Thread.sleep(3 * 1000); }
            catch(InterruptedException ignored) {}
            return 42;
        });
    }

    public ReceiveFuture(ActorContext<Integer> context) {
        super(context);

        System.out.println("Starting...");
        context.pipeToSelf(calculate(), (o, e) -> e == null ? o : -1);
    }

    public static Behavior<Integer> create() {
        return Behaviors.setup(ReceiveFuture::new);
    }

    public static void main(String[] args) {
        ActorSystem.create(ReceiveFuture.create(), "receive-future");
    }

    @Override
    public Receive<Integer> createReceive() {
        return newReceiveBuilder().onMessage(Integer.class, (message) -> {
            System.out.println(message);
            return Behaviors.stopped();
        }).build();
    }
}
